be/src/agent/publish_version.cpp

```diff
@@ -0,0 +1,161 @@
+// This file is licensed under the Elastic License 2.0. Copyright 2021-present, StarRocks Limited.
+
+#pragma once
+
+#include "publish_version.h"
+
+#include <bvar/bvar.h>
+
+#include "gutil/strings/join.h"
+#include "storage/data_dir.h"
+#include "storage/storage_engine.h"
+#include "storage/tablet.h"
+#include "util/monotime.h"
+#include "util/threadpool.h"
+
+namespace starrocks {
+
+const uint32_t PUBLISH_VERSION_SUBMIT_MAX_RETRY = 10;
+bvar::LatencyRecorder g_publish_latency("be", "publish");
+
+void flush_dirs(std::unordered_set<DataDir*>& affected_dirs) {
+    for (auto dir : affected_dirs) {
+        auto st = dir->get_meta()->flush();
+        if (!st.ok()) {
+            LOG(WARNING) << "failed to flush tablet meta, dir:" << dir->path() << " res:" << st;
+        }
+    }
+}
+
+struct TabletPublishVersionTask {
+    int64_t txn_id{0};
+    int64_t partition_id{0};
+    int64_t tablet_id{0};
+    int64_t version{0};
+    RowsetSharedPtr rowset;
+    Status st;
+    int64_t max_continuous_version{0}; // max continuous version after publish is done
+};
+
+void run_publish_version_task(ThreadPool& threadpool, const TAgentTaskRequest& publish_version_task,
+                              TFinishTaskRequest& finish_task, std::unordered_set<DataDir*>& affected_dirs) {
+    int64_t start_ts = MonotonicMillis();
+    auto publish_version_req = publish_version_task.publish_version_req;
```

Comment:
```suggestion
    auto& publish_version_req = publish_version_task.publish_version_req;
```
use reference to avoid struct copy